### PR TITLE
Port PackedSequences functions to C++

### PR DESCRIPTION
**Pull request by apaszke**
@zdevito

### CODE REVIEW

File: aten/src/ATen/native/PackedSequence.cpp [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+  AT_CHECK(lengths_t.size(0) == batch_size,
+           "Expected `len(lengths)` to be equal to batch_size, but got ", lengths_t.size(0),
+           " (batch_size=", batch_size, ")");
+  AT_CHECK(lengths[batch_size - 1] > 0,
+           "Length of all samples has to be greater than 0, but found an element "
+           "in 'lengths' that is <= 0");
+
+  std::vector<at::Tensor> steps;
+  steps.reserve(batch_size);
+  at::Tensor batch_sizes_t = _lengths.type().tensor(lengths[0]);
```
Comment (zdevito) [inline aten/src/ATen/native/PackedSequence.cpp]: I think this should be: `at::empty(lengths[0], _lengths.options());`

**Review by zdevito — APPROVED** [review]
This looks good -- I didn't check the indexing very closely, but the structure looks good, and it removes a good bit of onnx hacks.

**Review by jamesr66a — APPROVED** [review]
ONNX copypasta LGTM

**Pull request closed without merging**